### ISSUE TITLE
Can't create unique index on same attribute that is used for sharding

### ISSUE THREAD

**Issue by Kent767** [arangodb/arangodb#2131]
## my environment running ArangoDB

I'm using the latest ArangoDB of the respective release series:
- [x] 3.0

On this operating system:
- [X] DCOS on
  - [X] Azure

Repro Steps:
Create a new Collection, "TestCollection"
3 shards
ShardedBy: "xyz"

on that collection Create a unique index on "xyz" (doesn't seem to matter if it's skiplist/persistent/hash)

Expected Result:
  A unique index on xyz is created
Actual Result:
  The web ui closes the modal and seems to indicate everything went fine, but no index appears to be created.

**Comment by Kent767** [comment]
Also of note, you can't create a composite unique index which uses this attribute for one of it's fields.
Unique index on (xyz, and abc)

**Comment by dothebart** [comment]
In clusters unique indeces have to have additional attributes; they have to be in use to distribute the documents into the shards. - thus they have to be a `shard key`.
Have a look at:
https://docs.arangodb.com/devel/Manual/Administration/Sharding/index.html
for more details.

**Comment by jsteemann** [comment]
Creating a unique index on the shard key is supported.
The following works, for example:
```
db._create("TestCollection", { numberOfShards: 3, shardKeys: [ "xyz" ] }); db.TestCollection.ensureIndex({ type: "hash", unique: true, fields: ["xyz"] });
```

Creating a unique index on a non-shard key won't work however:
```
db._create("TestCollection", { numberOfShards: 3, shardKeys: [ "xyz" ] }); db.TestCollection.ensureIndex({ type: "hash", unique: true, fields: ["abc"] });

JavaScript exception in file '/home/steemann/Arango30NoAsan/js/client/modules/@arangodb/arangosh.js' at 100,7: ArangoError 1470: unsupported operation or parameter
```

It is also possible to create a composite unique index provided that the index attributes are identical to the shard keys:
```
db._create("TestCollection", { numberOfShards: 3, shardKeys: [ "abc", "xyz" ] }); db.TestCollection.ensureIndex({ type: "hash", unique: true, fields: ["abc", "xyz"] });
```

Again, if the index attributes do not match the shard keys, then the operation will fail:
```
db._create("TestCollection", { numberOfShards: 3, shardKeys: [ "abc", "xyz" ] }); db.TestCollection.ensureIndex({ type: "hash", unique: true, fields: ["abc"] });
JavaScript exception in file '/home/steemann/Arango30NoAsan/js/client/modules/@arangodb/arangosh.js' at 100,7: ArangoError 1470: unsupported operation or parameter
```

All the above is intentional as @dothebart posted. However, two issues remain:
* the web interface does not seem to use the shard key entered when creating a collection. When I created a collection with shard key "xyz", the shard key "_key" was used instead!
* no error is displayed in the web interface when index creation fails

**Comment by jsteemann** [comment]
The shard keys are now properly used as entered when creating a new collection via the web interface. The bug only existed in 3.0, but it worked in devel already.

The other issue (no error shown in web interface when index creation fails) is not yet fixed.

**Comment by Kent767** [comment]
Glad to know the webui was mostly the issue, I do have a followup question, however, given the other information that all unique indexes must be the shard key as well.  In this case I have data that makes sense to group together by, for example: "ClientAccount", all entities will have a ClientAccount attribute, that's not unique.  However the tuple on each entity: [ClientAccount,OriginalId] is unique from where we're sourcing the data.  In this case, setting shard_key to ClientAccount,OriginalId would result in presumably that client data getting distributed all over the cluster, (which may be undesirable given many queries may focus on 1 client account at a time)

what's the best practice for this, if I can't support doing ShardBy ClientAccount, and a unique Key on [ClientAccount,OriginalId] ?

Thanks again!

**Comment by jsteemann** [comment]
At the moment it's simply not possible to create a unique index on [ `ClientAccount`, `OriginalId` ] while using just `ClientAccount` as the shard key. However, I think this is over-restrictive and it should be allowed to create such unique indexes. I have created a patch (https://github.com/arangodb/arangodb/commit/2a371cbf9e7ce04b4a7be40b5e73dced12e52c31) that allows creating such indexes, and the tests indicate that it should work. 
I am now waiting for my colleagues' feedback on this and if nobody objects, I hope the patch can be merged into a release.

**Comment by Kent767** [comment]
Thanks! hopefully it's prioritized, I think lots of people should find that useful.

**Comment by jsteemann** [comment]
@Kent767 : the patch has been pulled into the devel branch already.
I think it will be merged into the 3.1 branch soon, and probably 3.0 as well.

**Comment by hkernbach** [comment]
@jsteemann: no error is displayed in the web interface when index creation fails

is done. added to 3.0/3.1 and devel.

**Comment by hkernbach** [comment]
ArangoDB 3.1 has been released. Please note that the download urls and repositories have changed slightly. 

[Download](https://www.arangodb.com/download) or see [Release Notes](https://docs.arangodb.com/3.1/Manual/ReleaseNotes/NewFeatures31.html) for details.

**Comment by jsteemann** [comment]
@Kent767 : the patch that will allow creating unique indexes on the shard keys plus additional attributes is still pending. It is not yet in the 3.1 branch, but I think we'll discuss the patch this week so it can eventually be merged into a release.

**Comment by jsteemann** [comment]
@Kent767 : just backported the fix into the 3.1 branch. Will be part of the next 3.1 release.
